src/Raven.Server/Documents/Sharding/Background/ShardedPeriodicDocumentsMigrator.cs

```diff
@@ -0,0 +1,106 @@
+        {
+            _database = database;
+        }
+
+        protected override async Task DoWork()
+        {
+            await WaitOrThrowOperationCanceled(_database.Configuration.Sharding.PeriodicDocumentsMigrationInterval.AsTimeSpan);
+
+            await ExecuteMoveDocuments();
+        }
+
+        internal async Task ExecuteMoveDocuments()
+        {
+            try
+            {
+                var buckets = new Dictionary<int, int>();
+                using (_database.ShardedDocumentsStorage.ContextPool.AllocateOperationContext(out DocumentsOperationContext context))
+                {
+                    while (true)
+                    {
+                        buckets.Clear();
+                        context.Reset();
+                        context.Renew();
+
+                        using (context.OpenReadTransaction())
+                        {
+                            var configuration = _database.ShardingConfiguration;
+                            for (var index = 0; index < configuration.BucketRanges.Count; index++)
+                            {
+                                var range = configuration.BucketRanges[index];
+                                if (range.ShardNumber == _database.ShardNumber)
+                                    continue;
+
+                                var start = range.BucketRangeStart;
+                                var end = index == configuration.BucketRanges.Count - 1
+                                    ? int.MaxValue
+                                    : configuration.BucketRanges[index + 1].BucketRangeStart;
+
+                                var bucketStatistics = ShardedDocumentsStorage.GetBucketStatistics(context, start, end);
+
+                                if (bucketStatistics == null)
+                                    continue;
+
+                                foreach (var bucketStats in bucketStatistics)
+                                    buckets.Add(bucketStats.Bucket, range.ShardNumber);
+                            }
+                        }
+
+                        if (buckets.Count > 0)
+                        {
+                            await MoveDocumentsToShard(buckets);
+                            return;
+                        }
+                    }
+                }
+            }
+            catch (OperationCanceledException)
+            {
+                // this will stop processing
+                throw;
+            }
+            catch (Exception e)
+            {
+                if (Logger.IsOperationsEnabled)
+                    Logger.Operations($"Failed to execute documents migration for '{_database.Name}'", e);
+
+                throw;
+            }
+        }
+
+        private async Task MoveDocumentsToShard(Dictionary<int, int> buckets)
```

Comment:
since we are doing one bucket at a time, lets avoid all the complexities here and have as a parameter only `int bucket`.

Reply:
done